firedrake/preconditioners/asm.py

```diff
@@ -103,6 +103,30 @@ def get_patches(self, V):
 
     def view(self, pc, viewer=None):
         self.asmpc.view(viewer=viewer)
+        self.prefix = pc.getOptionsPrefix() + self._prefix
+
+        if viewer is not None:
+            opts = PETSc.Options(self.prefix)
+            print_statistics = opts.getBool("print_patch_statistics", default=False)
+
+            if print_statistics:
+                from mpi4py import MPI
+
+                dm = pc.getDM()
+                V = get_function_space(dm)
+
+                # Obtain patches from user defined function
+                ises = self.get_patches(V)
```

Comment:
It feels slightly extravagant to completely recompute the patch indices just for the sake of this monitor, but it is true that we don't stash these anywhere convenient for us to find.  Is it worth stashing them in `__init__` so that we don't need to recompute them here?

Reply:
I agree it's a bit wasteful. I agree we could stash them in `__init__` if the option is present. I'll make this change.

Reply:
Isn't the correct approach here to compute the statistics during `__init__` if the options are present? Then just stash the statistics to save memory.